// Homework 2
// facebook.proto
// Server Protocol Buffers
// Colin Banigan and Katherine Drake

// use grpc 3
syntax = "proto3";

package facebookChat;

service fbChatRoom {
    // user runs client
    rpc Login(LoginRequest) returns (LoginReply) {}
    
    // LIST command
    rpc List(ListRequest) returns (ListReply) {}
    
    // JOIN command
    rpc Join(JoinRequest) returns (JoinReply) {}
    
    // LEAVE command
    rpc Leave(LeaveRequest) returns (LeaveReply) {}
    
    // CHAT command
    rpc Chat(ChatRequest) returns (ChatReply) {}
    
    // client sends chat message to server
    rpc SendChatToServer(ChatMessage) returns (ChatReply) {}
    
    // server sends chat message to client
    rpc SendChatToClient(ChatRequest) returns (ChatReply) {}
}

// Login Messages
message LoginRequest {
    string username = 1;
}
message LoginReply {
    string reply = 1;
}

// List Messages
message ListRequest {
    string username = 1;
}
message ListReply {
    string reply = 1;
    repeated string username = 2;
}

// Join Messages
message JoinRequest {
    string username = 1;
    string chatroom = 2;
}
message JoinReply {
    string reply = 1;
}

// Leave Messages
message LeaveRequest {
    string username = 1;
    string chatroom = 2;
}
message LeaveReply {
    string reply = 1;
}

// Chat Messages
message ChatRequest {
    string username = 1;
}
message ChatReply {
    string message = 1;
}
message ChatMessage {
    string username = 1;
    string message = 2;
}